Refactor functions worker runtime retrieval to use `IWorkerRuntimeResolver` abstraction

kshyju opened the pull request:
This PR introduces a new abstraction, `IWorkerRuntimeResolver`, for retrieving the Functions worker runtime value. All application code paths that previously read the runtime directly from environment variables have been updated to use this abstraction instead.

There are 2 implementations of this.

#### ScriptHostWorkerRuntimeResolver
Provides the worker runtime for the current Script Host scope.

- Applies custom logic when a configuration profile is present.
- Falls back to reading the value from the environment variable when no special handling is required.
- Caches the resolved value and returns it on subsequent calls.

#### WebHostWorkerRuntimeResolverAdapter
A Web Host–level adapter that:

- Delegates to `ScriptHostWorkerRuntimeResolver` when it becomes available.
- Falls back to reading from the environment variable until the Script Host–specific resolver can be used.

* [x] Backporting to the `in-proc` branch is not required
    * Otherwise: Link to backporting PR
* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)

fabiocav committed:
Initial set of changes to improve resolution of FUNCTIONS_WORKER_RUNTIME

kshyju committed:
Using IWorkerRuntimeResolver.GetWorkerRuntime() to retrieve functions worker runtime value instead of reading from environment.

kshyju committed:
release notes.

kshyju commented on src/WebJobs.Script/Host/HostFunctionMetadataProvider.cs:
```diff
@@ -104,6 +104,7 @@ private FunctionMetadata ReadFunctionMetadata(string functionDirectory, IFileSys
 
                     JObject functionConfig = JObject.Parse(json);
 
+                    // FACAVAL|FWR: Dependency on parent scope
```
Will follow up with Fabio to understand he had any specific changes here in mind.

kshyju replied:
Replaced env variable access with new abstraction.

kshyju committed:
Copilot review feedback fixes.

kshyju commented on src/WebJobs.Script/Workers/ScriptHostWorkerRuntimeResolver.cs:
```diff
@@ -0,0 +1,49 @@
+
+            if (string.Equals(_scriptJobHostOptionsMonitor.CurrentValue.ConfigurationProfile, "web-app-custom-handler"))
+            {
+                return _resolvedWorkerRuntime = "custom";
```
Updated to use OrdinalIgnoreCase comparison.

kshyju commented on src/WebJobs.Script.Grpc/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:
```diff
@@ -95,7 +97,8 @@ public RpcFunctionInvocationDispatcher(IOptions<ScriptJobHostOptions> scriptHost
             _logger = loggerFactory.CreateLogger<RpcFunctionInvocationDispatcher>();
             _rpcWorkerChannelFactory = rpcWorkerChannelFactory;
-            _workerRuntime = _environment.GetEnvironmentVariable(EnvironmentSettingNames.FunctionWorkerRuntime);
+            _workerRunimeResolver = workerRunimeResolver;
```
Local assignment was not needed. Removed now.

kshyju commented on test/WebJobs.Script.Tests/Workers/Rpc/RpcWorkerConfigTests.cs:
```diff
@@ -686,13 +686,14 @@ private IEnumerable<RpcWorkerConfig> TestReadWorkerProviderFromConfig(IEnumerabl
                 var scriptSettingsManager = new ScriptSettingsManager(config);
                 var workerProfileManager = new Mock<IWorkerProfileManager>();
                 var testMetricLogger = new TestMetricsLogger();
+                var workerRuntimeResolver = new Mock<IWorkerRuntimeResolver>(MockBehavior.Strict);
```
Removed.

kshyju commented on src/WebJobs.Script.Grpc/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:
```diff
@@ -95,7 +97,8 @@ public RpcFunctionInvocationDispatcher(IOptions<ScriptJobHostOptions> scriptHost
             _rpcWorkerChannelFactory = rpcWorkerChannelFactory;
-            _workerRuntime = _environment.GetEnvironmentVariable(EnvironmentSettingNames.FunctionWorkerRuntime);
+            _workerRunimeResolver = workerRunimeResolver;
+            _workerRuntime = _workerRunimeResolver.GetWorkerRuntime();
```
local variable is not needed anymore. Removed.

kshyju commented on test/WebJobs.Script.Tests/WebJobsScriptHostServiceTests.cs:
```diff
@@ -40,6 +41,7 @@ public class WebJobsScriptHostServiceTests
         private ILoggerFactory _loggerFactory;
         private Mock<IScriptWebHostEnvironment> _mockScriptWebHostEnvironment;
         private Mock<IEnvironment> _mockEnvironment;
+        private Mock<IWorkerRuntimeResolver> _mockWorkerRuntimeResolver;
```
Fixed.

kshyju commented on src/WebJobs.Script.WebHost/DependencyInjection/WebHostWorkerRuntimeResolverAdapter.cs:
```diff
@@ -0,0 +1,124 @@
+
+        public string GetWorkerRuntime(string defaultValue = null)
+        {
+            EnsureSubscribedToHostManagerStateChange();
+
+            var jobHostResolver = _cachedHostResolver;
+
+            if (jobHostResolver is null)
+            {
+                var scriptHostWorkerResolver = _rootProvider.GetScriptHostServiceOrNull<IWorkerRuntimeResolver>();
+                if (scriptHostWorkerResolver is not null)
+                {
+                    _logger.LogDebug("Script host worker resolver resolved successfully.");
+
+                    var existing = Interlocked.CompareExchange(
+                        ref _cachedHostResolver,
+                        scriptHostWorkerResolver,
+                        comparand: null);
+
+                    jobHostResolver = existing ?? scriptHostWorkerResolver;
+
+                    if (existing is null)
+                    {
+                        _logger.ScriptHostWorkerResolverCached();
+                    }
+                }
+            }
+
+            if (jobHostResolver is not null)
+            {
+                return jobHostResolver.GetWorkerRuntime(defaultValue);
+            }
+
+            // Fallback to read from environment when Job Host scoped resolver is not available yet
+            _environment ??= _rootProvider.GetRequiredService<IEnvironment>();
+
+            return _environment.GetEnvironmentVariableOrDefault(
+                EnvironmentSettingNames.FunctionWorkerRuntime,
+                defaultValue);
+        }
```
This singleton is disposed only during application shutdown. At that point, `GetWorkerRuntime()` is not expected to be called. So checking disposed in GetWorkerRuntime is unnecessary.

kshyju committed:
Fixed one code path where I missed to replace environment variable access with the new abstraction method.

kshyju committed:
Fixed a typo.

kshyju commented on src/WebJobs.Script.Grpc/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:
```diff
@@ -78,7 +79,8 @@ public RpcFunctionInvocationDispatcher(IOptions<ScriptJobHostOptions> scriptHost
             IOptions<FunctionsHostingConfigOptions> hostingConfigOptions,
-            IHostMetrics hostMetrics)
+            IHostMetrics hostMetrics,
+            IWorkerRuntimeResolver workerRunimeResolver)
```
Fixed typo.

kshyju committed:
Add thread safety when setting cache,

kshyju committed:
Thread safety improvements in Dispose.

kshyju commented on src/WebJobs.Script.WebHost/DependencyInjection/WebHostWorkerRuntimeResolverAdapter.cs:
```diff
@@ -0,0 +1,124 @@
+        private IWorkerRuntimeResolver _cachedHostResolver;
+        private IScriptHostManager _hostManager;
+        private IEnvironment _environment;
+        private bool _disposed;
```
using of disposed is in a thread safe manner now.

kshyju commented on src/WebJobs.Script/Workers/ScriptHostWorkerRuntimeResolver.cs:
```diff
@@ -0,0 +1,50 @@
+{
+    /// <summary>
+    /// Resolves the worker runtime for the current script host instance.
+    /// </summary>
+    internal sealed class ScriptHostWorkerRuntimeResolver : IWorkerRuntimeResolver
+    {
+        private readonly IEnvironment _environment;
+        private readonly IOptionsMonitor<ScriptJobHostOptions> _scriptJobHostOptionsMonitor;
+        private string _resolvedWorkerRuntime;
+
+        public ScriptHostWorkerRuntimeResolver(
+            IEnvironment environment,
+            IOptionsMonitor<ScriptJobHostOptions> scriptJobHostOptionsMonitor)
+        {
+            ArgumentNullException.ThrowIfNull(environment);
+            ArgumentNullException.ThrowIfNull(scriptJobHostOptionsMonitor);
+
+            _environment = environment;
+            _scriptJobHostOptionsMonitor = scriptJobHostOptionsMonitor;
+        }
+
+        public string GetWorkerRuntime(string defaultValue = null)
+        {
+            if (_resolvedWorkerRuntime is not null)
+            {
+                return _resolvedWorkerRuntime;
+            }
+
+            if (string.Equals(_scriptJobHostOptionsMonitor.CurrentValue.ConfigurationProfile,
+                              HostConfigurationProfile.McpCustomHandlerProfile,
+                              StringComparison.OrdinalIgnoreCase)
+                || string.Equals(_scriptJobHostOptionsMonitor.CurrentValue.ConfigurationProfile,
+                                 HostConfigurationProfile.WebAppCustomHandlerProfile,
+                                 StringComparison.OrdinalIgnoreCase))
+            {
+                return _resolvedWorkerRuntime = "custom";
+            }
+
+            return _resolvedWorkerRuntime = _environment.GetEnvironmentVariableOrDefault(EnvironmentSettingNames.FunctionWorkerRuntime, defaultValue);
+        }
```
_resolvedWorkerRuntime is thead safe now.

kshyju committed:
Missed another instance of direct enviornment access to retrieve FWR.

kshyju commented on src/WebJobs.Script/Workers/ScriptHostWorkerRuntimeResolver.cs:
```diff
@@ -0,0 +1,58 @@
+            else
+            {
+                workerRuntime = _environment.GetEnvironmentVariableOrDefault(EnvironmentSettingNames.FunctionWorkerRuntime, defaultValue);
+            }
```
No. I prefer the existing approach of style in this case. I find that to be equally readable.

kshyju committed:
Switch to src gen logging.

brettsam commented on src/WebJobs.Script.WebHost/DependencyInjection/WebHostWorkerRuntimeResolverAdapter.cs:
```diff
@@ -0,0 +1,122 @@
+        private readonly ILogger<WebHostWorkerRuntimeResolverAdapter> _logger;
+        private IWorkerRuntimeResolver _cachedHostResolver;
+        private IScriptHostManager _hostManager;
+        private IEnvironment _environment;
```
We're generally trying to reduce our usage of `IEnvironment` -- can this use `IConfiguration` instead?

brettsam commented on src/WebJobs.Script.WebHost/DependencyInjection/WebHostWorkerRuntimeResolverAdapter.cs:
```diff
@@ -0,0 +1,122 @@
+            }
+
+            // Fallback to read from environment when Job Host scoped resolver is not available yet
+            _environment ??= _rootProvider.GetRequiredService<IEnvironment>();
```
Why would `_rootProvider` need to be called here? I thought the passed in `IEnvironment` already would have come from the root?

kshyju replied:
Fixed.

brettsam commented on src/WebJobs.Script.WebHost/DependencyInjection/WebHostWorkerRuntimeResolverAdapter.cs:
```diff
@@ -0,0 +1,122 @@
+                return;
+            }
+
+            hostManager = _rootProvider.GetRequiredService<IScriptHostManager>();
```
Same question about the root provider here -- or maybe I'm not quite understanding the scope of these...

brettsam replied:
Want to bump this up -- can't we inject IScriptHostManager into this class and not have to do the service locator thing?

We may have talked about it already but I forget :-)

brettsam commented on src/WebJobs.Script.WebHost/DependencyInjection/WebHostWorkerRuntimeResolverAdapter.cs:
```diff
@@ -0,0 +1,122 @@
+                var scriptHostWorkerResolver = _rootProvider.GetScriptHostServiceOrNull<IWorkerRuntimeResolver>();
+                if (scriptHostWorkerResolver is not null)
+                {
+                    _logger.ScriptHostWorkerResolverResolvedSuccessfully();
```
How noisy is this? Want to make sure we're not flooding Kusto...

liliankasem replied:
+1

kshyju replied:
Expect this to be logged once per script host instance.

brettsam commented on src/WebJobs.Script/Workers/ScriptHostWorkerRuntimeResolver.cs:
```diff
@@ -0,0 +1,58 @@
+    /// <summary>
+    /// Resolves the worker runtime for the current script host instance.
+    /// </summary>
+    internal sealed class ScriptHostWorkerRuntimeResolver : IWorkerRuntimeResolver
```
Did you look into having this be part of an Options object instead of our own interface? On the surface it looks like it could work as we want to store/cache a simple string at the end of the day and a ConfigureOptions class could do all of that...

liliankasem replied:
I got the same AI suggestion for this - @kshyju did you look into IOptions instead?

>The .NET Options pattern already has built-in caching, change-token support, and DI integration. An alternative to IWorkerRuntimeResolver would be a WorkerRuntimeOptions class with a Runtime property, configured via IConfigureOptions<WorkerRuntimeOptions> (the setup reads the env var / profile). Consumers would inject IOptionsMonitor<WorkerRuntimeOptions> and get automatic change notification and caching from the framework. This would eliminate the hand-rolled Interlocked.CompareExchange caching and ActiveHostChanged subscription logic, relying instead on the same pattern already used for LanguageWorkerOptions.

kshyju replied:
Yes, the IOptions approach was considered. The key challenge is that this resolver bridges two DI scopes. The WebHost singleton scope and the Script Host child scope, which the Options pattern doesn't handle well out of the box:

Cross-scope delegation: WebHostWorkerRuntimeResolverAdapter (WebHost singleton) needs to delegate to the Script Host-scoped resolver when available, falling back to environment variable reads when no host is running. IOptionsMonitor<T> doesn't have a concept of "delegate to a child scope's options when available."

Specialization timing: During placeholder mode, the env var isn't set yet. After specialization, it is. The adapter must re-read on each call until a value appears, then cache it. IConfigureOptions runs once at options creation time & it wouldn't naturally re-evaluate on each access.

The IWorkerRuntimeResolver abstraction gives us explicit control over these scope-bridging and lifecycle concerns. That said, if we find the hand-rolled caching becomes a maintenance burden, we could revisit. The ScriptHostWorkerRuntimeResolver (the Script Host-scoped one) is simpler and could potentially be an Options object, but keeping both implementations behind the same interface is cleaner for consumers.

@brettsam  agreed it is not a bad idea to use custom abstraction instead of Options here.

kshyju committed:
Use Configuration in WebHostWorkerRuntimeResolverAdapter

kshyju committed:
formatting fixes (whitespace removal)

kshyju committed:
Revert back IEnvironment

brettsam commented:
@kshyju -- did you want to get these merges cleaned up and then we can continue the review? I'm okay with this approach now that I saw how complicated some of the IOptions<> dependencies can get...

kshyju committed:
clean(fix merge conflict)

kshyju committed:
Manual merge coflict fix after rebase.

kshyju committed:
PR feedback fixes

kshyju committed:
remove caching of env var value when it is null

kshyju committed:
Merge branch 'dev' into fabiocav/fwrresolver

kshyju committed:
Fix line break.

kshyju committed:
more linting fixes.

kshyju committed:
GHCP PR feedback fixes.

kshyju committed:
Fix order of using statements

kshyju committed:
Update WebJobsScriptHostService logging code to read from env for original runtime

kshyju committed:
Minor cleanup

kshyju committed:
Do not cache default value.

kshyju commented on test/WebJobs.Script.Tests/Workers/Rpc/WorkerConfigurationProviderBaseTests.cs:
```diff
@@ -158,7 +167,7 @@ public void WorkerDescription_Skipped_When_Profile_Disables_Worker()
             var loggerFactory = GetTestLoggerFactory();
             var testScriptHostManager = new Mock<IScriptHostManager>();
-            var optionsMonitor = GetTestWorkerConfigurationResolverOptions(config, testEnvironment, testScriptHostManager.Object);
+            var optionsMonitor = GetTestWorkerConfigurationResolverOptions(config, workerRuntime: null, testEnvironment, testScriptHostManager.Object);
```
This is a false positive. Since C# 7.2, named arguments are permitted in non-trailing positions as long as they are in the correct ordinal position ([spec reference](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-version-history#c-version-72)). In all of these call sites, `workerRuntime:` is the second argument matching the second parameter in the `GetTestWorkerConfigurationResolverOptions` signature, so positional arguments can legally follow it. The project builds without errors or warnings.

kshyju commented on test/WebJobs.Script.Tests/Workers/Rpc/DynamicWorkerConfigurationProviderTests.cs:
```diff
@@ -57,7 +55,12 @@ public void GetWorkerConfigs_MultiLanguageWorker_ReturnsExpectedConfigs(string r
             hostingOptions.Features.Add(RpcWorkerConstants.WorkersAvailableForDynamicResolution, "java|node");
-            var optionsMonitor = GetTestWorkerConfigurationResolverOptions(config, mockEnvironment.Object, testScriptHostManager.Object, new OptionsWrapper<FunctionsHostingConfigOptions>(hostingOptions));
+            var optionsMonitor = GetTestWorkerConfigurationResolverOptions(
+                config,
+                workerRuntime: null,
+                mockEnvironment.Object,
+                testScriptHostManager.Object,
+                new OptionsWrapper<FunctionsHostingConfigOptions>(hostingOptions));
```
This is a false positive. Since C# 7.2, named arguments are permitted in non-trailing positions as long as they are in the correct ordinal position ([spec reference](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-version-history#c-version-72)). In all of these call sites, `workerRuntime:` is the second argument matching the second parameter in the `GetTestWorkerConfigurationResolverOptions` signature, so positional arguments can legally follow it. The project builds without errors or warnings.

kshyju commented:
> @kshyju -- did you want to get these merges cleaned up and then we can continue the review? I'm okay with this approach now that I saw how complicated some of the IOptions<> dependencies can get...

@brettsam Rebased and ready to review.

liliankasem commented on src/WebJobs.Script/Host/ScriptHost.cs:
```diff
@@ -471,9 +477,9 @@ private void PreInitialize()
+        internal static void ValidateFunctionsWorkerRuntime(IWorkerRuntimeResolver workerRuntimeResolver, IOptions<FunctionsHostingConfigOptions> hostingConfigOptions, ILogger logger)
         {
-            if (string.IsNullOrEmpty(environment.GetFunctionsWorkerRuntime()))
+            if (string.IsNullOrEmpty(workerRuntimeResolver.GetWorkerRuntime()))
```
Can't this just use the global resolver (`_workerRuntimeResolver`)? Why do we need to pass it as a param?

kshyju replied:
The method is internal static so that it can be unit tested directly in ScriptHostTests.cs without constructing a full ScriptHost instance.

liliankasem commented on src/WebJobs.Script/Workers/IWorkerRuntimeResolver.cs:
```diff
@@ -0,0 +1,20 @@
+    /// <summary>
+    /// Defines a mechanism for resolving the functions worker runtime value for the current context.
+    /// </summary>
+    public interface IWorkerRuntimeResolver
```
Could this be internal?

kshyju replied:
There are public class constructors to which this is being injected. So this interface has to be also public.

liliankasem commented on src/WebJobs.Script/Workers/ScriptHostWorkerRuntimeResolver.cs:
```diff
@@ -0,0 +1,69 @@
+
+        public string GetWorkerRuntime(string defaultValue = null)
+        {
+            var cachedRuntime = _resolvedWorkerRuntime;
+            if (cachedRuntime is not null)
+            {
+                return cachedRuntime;
+            }
```
no need for extra allocation here?

```suggestion
            if (_resolvedWorkerRuntime is not null)
            {
                return _resolvedWorkerRuntime;
            }
```

kshyju replied:
The local variable copies only the reference (a pointer-sized value on the stack), not the string contents, so there is no heap allocation involved. The reason it exists is thread safety. `_resolvedWorkerRuntime` is a shared field . Without the local, the JIT/CPU could issue two separate reads of the field, one for the null check and one for the return, and those reads could observe different values. Capturing a single snapshot in a local guarantees the null check and the return operate on the same value, eliminating a TOCTOU (time-of-check-time-of-use) race. This is a standard lock-free pattern.

liliankasem commented on src/WebJobs.Script/Workers/ScriptHostWorkerRuntimeResolver.cs:
```diff
@@ -0,0 +1,69 @@
+            _scriptJobHostOptionsMonitor = scriptJobHostOptionsMonitor;
+        }
+
+        public string GetWorkerRuntime(string defaultValue = null)
```
nit: might be nice to have a convenience property i.e.

`public string WorkerRuntime => GetWorkerRuntime();`

kshyju replied:
We don't want anyone to directly use this class. All access should be via the new abstraction (`IWorkerRuntimeResolver.GetWorkerRuntime`). Is there a use case you have in mind where exposing this property is helpful?

liliankasem commented on src/WebJobs.Script/Config/HostConfigurationProfile.cs:
```diff
@@ -29,6 +25,10 @@ public sealed class HostConfigurationProfile
 
+        internal const string McpCustomHandlerProfile = "mcp-custom-handler";
+
+        internal const string WebAppCustomHandlerProfile = "web-app-custom-handler";
```
These were propbably private since only this class used them, instead of making them internal - does it make more sense to move these to a Constants file?

kshyju replied:
_Agent response, which I agree._

They were made internal because ScriptHostWorkerRuntimeResolver now needs to compare against these profile names to determine the worker runtime . The constants are semantically tied to the profile concept — they define profile names and are used alongside HostConfigurationProfile.Get(), the profile static instances (McpCustomHandler, WebAppCustomHandler), and the SupportedValues string. Moving them to a generic Constants file would separate them from the type they describe. The existing comment at line 14 ("note: profile name consts are intentionally private...") already explains the original design intent — changing to internal still preserves the test-safety property since tests reference these constants too. Prefer keeping them in HostConfigurationProfile for cohesion.

liliankasem replied:
not a hill I want to die on :)

liliankasem commented on src/WebJobs.Script.Grpc/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:
```diff
@@ -198,7 +199,7 @@ internal async Task InitializeWebhostLanguageWorkerChannel(IEnumerable<string> l
         internal async void ShutdownWebhostLanguageWorkerChannels()
         {
-            _logger.LogDebug("{workerRuntimeConstant}={value}. Will shutdown all the worker channels that started in placeholder mode", EnvironmentSettingNames.FunctionWorkerRuntime, _workerRuntime);
+            _logger.LogDebug("Worker runtime={value}. Will shutdown all the worker channels that started in placeholder mode", _workerRuntime);
```
Will this impact any detectors?

kshyju replied:
I am not aware of any. But not 100% sure about it.  Changing this log statement here makes it correct (with this new changes).  What is the best way to check this?

liliankasem replied:
I don't know if there is an easy way - one approach is to open the detectors repo and search for the original log statement to see if it comes up at all

kshyju replied:
Checked for "Will shutdown all the worker channels" message in detector repo on main as of 3/18 and no entries found.

liliankasem commented on src/WebJobs.Script/Workers/ScriptHostWorkerRuntimeResolver.cs:
```diff
@@ -0,0 +1,69 @@
+            _scriptJobHostOptionsMonitor = scriptJobHostOptionsMonitor;
+        }
+
+        public string GetWorkerRuntime(string defaultValue = null)
```
Do we want to consider introducing an enum here instead of using string? Since we are already refactoring, it might be nice to avoid all the string == const checks

kshyju replied:
Did you mean changing `_scriptJobHostOptionsMonitor.CurrentValue.ConfigurationProfile` to enum instead of string? That might need additional changes on how the Option is populated etc. Perhaps a future improvement if we think it adds any value. This PR is already big, and I prefer to avoid increasing the scope further.

liliankasem replied:
I just meant that worker runtime should be an enum i.e.

```chsarp
    public enum WorkerRuntime
    {
        None,
        Dotnet,
        DotnetIsolated,
        Node,
        Python,
        Java,
        Powershell,
        Custom
    }
```

liliankasem replied:
Does not necessarily mean `_scriptJobHostOptionsMonitor.CurrentValue.ConfigurationProfile` needs to change, just what `GetWorkerRuntime` returns and how the consumers of this method check what the language is

kshyju replied:
Asked agent to prepare pros and cons and impact.

<blockquote>
  Impact of Switching Worker Runtime from string to enum

  What exists today

  7 known runtime values in RpcWorkerConstants:

   - dotnet, dotnet-isolated, node, java, powershell, python, custom

  These are compared via case-insensitive string.Equals in ~25+ locations across ~30 source files with ~300+ references to workerRuntime/_workerRuntime/WorkerRuntime.

  Pros

   1. Type safety — Eliminates misspelled runtime strings. Compiler catches invalid values.
   2. No more case-sensitivity bugs — Today there's inconsistency: some comparisons use OrdinalIgnoreCase, others InvariantCultureIgnoreCase, and one at RpcFunctionInvocationDispatcher:480 uses plain string.Equals with no StringComparison at all (an existing bug). An enum eliminates all of these.
   3. Cleaner branching — switch expressions on enum vs. chained string.Equals calls (e.g., WebHostRpcWorkerChannelManager:210-212 checks node/powershell/python in 3 separate string comparisons).
   4. Performance — Enum comparison is a single integer op vs. string comparison. Marginal, but it's on a hot path (resolver is called frequently).

  Cons / Risks

   1. Massive scope — ~300+ references across ~30 source files in src/, plus an equivalent or larger number in test files. This is not a small refactor.
   2. String↔Enum boundary — The runtime value originates as a string from the FUNCTIONS_WORKER_RUNTIME environment variable and flows through several string-based systems:
    - WorkerDescription.Language is a string property used throughout the worker config pipeline. WorkerConfigs are matched to the runtime via c.Description.Language.Equals(workerRuntime). Changing this to enum means touching the entire worker configuration resolution chain.
    - FunctionMetadata.Language is also a string. It's compared against the runtime in Utility.cs for function validation.
    - Logging — The runtime string is interpolated into 20+ log messages. Each would need .ToString() or a format change.
    - Dictionary keys — WebHostRpcWorkerChannelManager._workerChannels uses the runtime string as a dictionary key.
   3. Unknown/custom runtimes — The enum would need an Unknown or Other variant for unrecognized values from the environment variable. This loses information (the actual string) and requires special handling everywhere.
   4. Serialization concerns — The runtime string appears in configuration (host.json via HostJsonFileConfigurationSource:261), API responses (HostController.GetWorkerProcesses), and metrics/telemetry. All of these expect the string form.
   5. Worker profile system — WorkerDescriptionProfile has ProfileDescription.Language as a string that gets compared to the runtime. The profile system would need parallel changes.
   6. External contract — Worker descriptions loaded from worker.config.json files define Language as a string. The parsing pipeline would need enum conversion at the boundary.

  Bottom line

  The enum is a better domain model, but the runtime string is deeply embedded as a cross-cutting concern — it flows from environment variables through configuration, worker registration, function metadata, channel management, logging, and API responses. The conversion boundary (string → enum at input, enum →    
  string at output) would need to be introduced at multiple layers. This is a dedicated refactoring effort, not something to fold into the resolver PR.
</blockquote>

liliankasem replied:
I think we should do it since we are refactoring anyway, but OK for this to be out of scope of this PR

liliankasem commented on src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs:
```diff
@@ -82,10 +83,20 @@ public class WebJobsScriptHostService : IHostedService, IScriptHostManager, ISer
+            IHostMetrics hostMetrics,
+            IWorkerRuntimeResolver workerRuntimeResolver,
             IOptions<FunctionsHostingConfigOptions> hostingConfigOptions,
             WorkerConfigCacheInvalidator workerConfigCacheInvalidator)
```
Maybe out of scope for this PR, but is the workerConfigCacheInvalidator thread safe? A couple of suggestions came up from the AI review about this class - let me know if you think these concerns are valid:

---

#### Guard the WorkerConfigCacheInvalidator downcast with an assertion

In `WorkerConfigCacheInvalidator.InvalidateCache()`, the pattern match against `RefreshWorkerOptionsChangeTokenSource<T>` silently no-ops if a different IOptionsChangeTokenSource<T> is registered. Consider either:

1. Adding a `Debug.Assert()` or warning log when the downcast fails, so misconfiguration is detectable
2. Accepting `RefreshWorkerOptionsChangeTokenSource<T>` directly as the constructor parameter type instead of the interface, since the class explicitly depends on that concrete behavior (the xmldoc already says this, but the type system doesn't enforce it)

#### Make WorkerConfigCacheInvalidator thread-safety assumptions explicit

The `_firstRun` and `_postBuildInvalidationEnabled` fields are plain bool with no synchronization, relying on the caller holding the host start semaphore. This assumption is implicit. Options:

1. Add a code comment or xmldoc noting the expected threading model
2. Or use Interlocked operations on int flags (like the adapter's _disposed) for defense-in-depth, which is cheap and avoids a subtle bug if future callers bypass the semaphore

kshyju replied:
Good suggestions. Like you said, it should be a separate PR addressing that.

liliankasem commented on src/WebJobs.Script.WebHost/DependencyInjection/WebHostWorkerRuntimeResolverAdapter.cs:
```diff
@@ -0,0 +1,143 @@
+            _logger = logger;
+        }
+
+        public string GetWorkerRuntime(string defaultValue = null)
```
AI suggestion:

> The defaultValue parameter on GetWorkerRuntime(string defaultValue = null) prevents caching null results (since different callers pass different defaults). If instead the interface returned string? with no default parameter, callers would do resolver.GetWorkerRuntime() ?? "fallback" at the call site. This would let the resolver always cache its resolved value (including null = "no runtime configured"), improving performance on the pre-specialization path where every call currently re-reads the environment variable.

kshyju replied:
The current behavior to not cache the default value passed in is intentional. I have code comments in the method and a test to ensure that. This is to prevent cache poisoning. One caller can update the cache with the default value and another caller may get it before it gets  a chance to see /readthe actual underlying env value.

kshyju replied:
My copilot CLI agent review flagged this as a major problem and I then ended up fixing it.

liliankasem reviewed (commented):
initial review

kshyju committed:
Merge branch 'dev' into fabiocav/fwrresolver

kshyju commented on test/WebJobs.Script.Tests/Workers/Rpc/WorkerConfigurationProviderBaseTests.cs:
```diff
@@ -41,7 +41,7 @@ public void DefaultLanguageWorkersDir()
             var testLoggerFactory = GetTestLoggerFactory();
 
-            var optionsMonitor = GetTestWorkerConfigurationResolverOptions(config, testEnvironment, testScriptHostManager.Object, null);
+            var optionsMonitor = GetTestWorkerConfigurationResolverOptions(config, workerRuntime: null, testEnvironment, testScriptHostManager.Object, null);
```
false alarm.

brettsam commented on src/WebJobs.Script/Workers/ScriptHostWorkerRuntimeResolver.cs:
```diff
@@ -0,0 +1,69 @@
+        private string _resolvedWorkerRuntime;
+
+        public ScriptHostWorkerRuntimeResolver(
+            IEnvironment environment,
```
We've been talking about getting rid of `IEnvironment` and using `IConfiguration` everywhere -- can we take that step here and populate this from `IConfiguration` instead?

kshyju replied:
Switched to read from IConfiguration.

brettsam commented on src/WebJobs.Script/Workers/ScriptHostWorkerRuntimeResolver.cs:
```diff
@@ -0,0 +1,69 @@
+                                 HostConfigurationProfile.WebAppCustomHandlerProfile,
+                                 StringComparison.OrdinalIgnoreCase))
+            {
+                workerRuntime = RpcWorkerConstants.CustomHandlerLanguageWorkerName;
```
I don't love this being hardcoded in here -- it makes it tough to search through the codebase to see "what exactly does the mcp profile do?". If we do switch to `IConfiguration`, I wonder if it'd make more sense to have these profiles as "real" config providers and inject their settings into `IConfiguration` all in one spot?

I don't think it should block this, but I do see it as something we're trying to improve on as we change things.

kshyju replied:
Good suggestion. In the latest iteration, I removed the code to check profile information from this class and moved it to HostConfigurationProfile class where it will add to IConfiguration. (This is a new config entry called "ProfileWorkerRuntime" which will be used if present when FWR is requested, else fallback to our existing logic to read FUNCTIONS_WORKER_RUNTIME from IConfiguration).

brettsam commented on src/WebJobs.Script.WebHost/DependencyInjection/WebHostWorkerRuntimeResolverAdapter.cs:
```diff
@@ -0,0 +1,143 @@
+                return cachedValue;
+            }
+
+            var valueFromEnvironment = _environment.GetEnvironmentVariable(EnvironmentSettingNames.FunctionWorkerRuntime);
```
Same question as elsewhere... can we pull this from `IConfiguration` and phase out `IEnvironment`?

kshyju replied:
Switched to read from IConfiguration.

kshyju committed:
Use IConfiguration in worker runtime resolvers and centralize profile settings

   - Switch WebHostWorkerRuntimeResolverAdapter and ScriptHostWorkerRuntimeResolver
     to use IConfiguration instead of IEnvironment for resolving FUNCTIONS_WORKER_RUNTIME.
   - Move the hardcoded worker runtime value ("custom") for MCP and WebApp custom handler
     profiles into HostConfigurationProfile, making profile behavior discoverable in one place.
   - Add ProfileWorkerRuntime property to ScriptJobHostOptions, auto-bound from profile config.
   - Surface FUNCTIONS_WORKER_RUNTIME into the script host's IConfiguration via a targeted
     AddInMemoryCollection in ScriptHostBuilderExtensions, avoiding the risks of exposing
     all environment variables.

kshyju committed:
Fix for tests

kshyju commented on src/WebJobs.Script.WebHost/DependencyInjection/WebHostWorkerRuntimeResolverAdapter.cs:
```diff
@@ -0,0 +1,144 @@
+                return;
+            }
+
+            hostManager = _rootProvider.GetRequiredService<IScriptHostManager>();
```
This is intentional. We cannot DI to constructor as that would cause a circular dependency. 

   IWorkerRuntimeResolver (WebHostWorkerRuntimeResolverAdapter)
     └─ needs IScriptHostManager (WebJobsScriptHostService)
          └─ needs IWorkerRuntimeResolver  ← CIRCULAR

So we are doing a deferred resolution here.

brettsam commented on src/WebJobs.Script/Workers/ScriptHostWorkerRuntimeResolver.cs:
```diff
@@ -40,19 +39,13 @@ public string GetWorkerRuntime(string defaultValue = null)
+            if (string.IsNullOrEmpty(workerRuntime))
             {
-                workerRuntime = _environment.GetEnvironmentVariable(EnvironmentSettingNames.FunctionWorkerRuntime);
+                workerRuntime = _configuration[EnvironmentSettingNames.FunctionWorkerRuntime];
```
Just double-checking... if this isn't present does it return null or throw?

kshyju replied:
It returns `null` if key is not present. Won't throw.

kshyju committed:
Merge branch 'dev' into fabiocav/fwrresolver

kshyju committed:
Merge branch 'dev' into fabiocav/fwrresolver

satvu commented on src/WebJobs.Script.Grpc/WorkerFunctionMetadataProvider.cs:
```diff
@@ -54,7 +58,7 @@ public ImmutableDictionary<string, ImmutableArray<string>> FunctionErrors
         public async Task<FunctionMetadataResult> GetFunctionMetadataAsync(IEnumerable<RpcWorkerConfig> workerConfigs, bool forceRefresh)
         {
-            _workerRuntime = _environment.GetEnvironmentVariable(EnvironmentSettingNames.FunctionWorkerRuntime);
+            _workerRuntime = _workerRuntimeResolver.GetWorkerRuntime();
```
Just curious - why is `_workerRuntime` re-calculated multiple times in this file (instead of using the stored value in constructor line 51)? Do we expect the value to change?

kshyju replied:
WorkerFunctionMetadataProvider is created in webhost scope. This method gets called after other life cycle events of script host such as specialization and we want to get the latest data at that time.
